File: src/main/resources/scripts/py/single.py
from timeout import init as timeOutInit
from loadbuilder import init as loadBuilderInit
#
from com.emc.mongoose.common.conf import RunTimeConfig
from com.emc.mongoose.common.logging import LogUtil
#
from com.emc.mongoose.core.api.io.task import IOTask
#
from org.apache.logging.log4j import Level, LogManager
#
from java.lang import Throwable, IllegalArgumentException, InterruptedException
from java.util import NoSuchElementException
#
LOG = LogManager.getLogger()
#
def init():
	loadBuilder = loadBuilderInit()
	try:
		loadType = IOTask.Type.valueOf(
			RunTimeConfig.getContext().getString(RunTimeConfig.KEY_SCENARIO_SINGLE_LOAD).upper()
		)
		LOG.debug(LogUtil.MSG, "Using load type: {}", loadType.name())
		loadBuilder.setLoadType(loadType)
	except NoSuchElementException:
		LOG.error(LogUtil.ERR, "No load type specified, try arg -Dscenario.single.load=<VALUE> to override")
	except IllegalArgumentException:
		LOG.error(LogUtil.ERR, "No such load type, it should be a constant from Load.Type enumeration")
	return loadBuilder
#
def build(loadBuilder):
	load = None
	if loadBuilder is None:
		LOG.fatal(LogUtil.ERR, "No load builder specified")
	else:
		try:
			load = loadBuilder.build()
		except Throwable as e:
			LogUtil.exception(LOG, Level.FATAL, e, "Failed to instantiate the load executor")
	return load
#
def execute(load):
	if load is None:
		LOG.fatal(LogUtil.ERR, "No load job specified")
	else:
		runTimeOut = timeOutInit()
		load.start()
		try:
			load.join(runTimeOut[1].toMillis(runTimeOut[0]))
		finally:
			load.close()
#
if __name__ == "__builtin__":
	loadBuilder = init()
	load = build(loadBuilder)
	try:
		execute(load)
	except InterruptedException as e:
		LOG.debug(LogUtil.MSG, "Single was interrupted")
	except Throwable as e:
		LogUtil.exception(LOG, Level.ERROR, e, "Scenario failed")
	loadBuilder.close() # to exit normally
	LOG.info(LogUtil.MSG, "Scenario end")

File: src/main/java/com/emc/mongoose/run/scenario/Single.java
package com.emc.mongoose.run.scenario;
//
import com.emc.mongoose.common.conf.RunTimeConfig;
import com.emc.mongoose.common.log.LogUtil;
import com.emc.mongoose.common.log.Markers;
//
import com.emc.mongoose.core.api.io.task.IOTask;
import com.emc.mongoose.core.api.load.builder.LoadBuilder;
import com.emc.mongoose.core.api.load.executor.LoadExecutor;
//
import com.emc.mongoose.util.builder.LoadBuilderFactory;
//
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
//
import java.io.IOException;
import java.rmi.RemoteException;
import java.util.concurrent.TimeUnit;
/**
 Created by kurila on 09.06.15.
 */
public final class Single
implements Runnable {
	//
	private final static Logger LOG;
	static {
		LogUtil.init();
		LOG = LogManager.getLogger();
	}
	//
	private LoadExecutor loadJob;
	private long timeOut;
	private TimeUnit timeUnit;
	//
	public Single(final AppConfig appConfig) {
		try(final LoadBuilder loadBuilder = LoadBuilderFactory.getInstance(appConfig)) {
			final IOTask.Type loadType = IOTask.Type.valueOf(
				appConfig.getString(RunTimeConfig.KEY_SCENARIO_SINGLE_LOAD).toUpperCase());
			LOG.debug(Markers.MSG, "Using load type: {}", loadType.name());
			loadBuilder.setLoadType(loadType);
			//
			final long timeOut = appConfig.getLoadLimitTimeValue();
			//
			this.timeOut = timeOut > 0 ? timeOut : Long.MAX_VALUE;
			this.timeUnit = timeOut > 0 ? appConfig.getLoadLimitTimeUnit() : TimeUnit.DAYS;
			this.loadJob = loadBuilder.build();
		} catch(final IOException e) {
			LogUtil.exception(LOG, Level.FATAL, e, "Failed to build the load job");
		}
	}
	//
	@Override
	public final void run() {
		try {
			loadJob.start();
			loadJob.await(timeOut, timeUnit);
		} catch(final RemoteException e) {
			LogUtil.exception(
				LOG, Level.WARN, e, "Failed to invoke a remote method for the load job: {}", loadJob
			);
		} catch(final InterruptedException e) {
			LOG.debug(Markers.MSG, "Load job \"{}\" was interrupted", loadJob);
		} finally {
			if(loadJob != null) {
				try {
					loadJob.close();
				} catch(final IOException e) {
					LogUtil.exception(
						LOG, Level.WARN, e, "Failed to close the load job: {}", loadJob
					);
				}
			}
		}
	}
	//
	public static void main(final String... args) {
		//
		try {
			RunTimeConfig.initContext();
			final AppConfig appConfig = BasicConfig.CONTEXT_CONFIG.get();
			//
			LOG.info(Markers.MSG, runTimeConfig);
			//
			final Single singleLoadScenario = new Single(runTimeConfig);
			singleLoadScenario.run();
			LOG.info(Markers.MSG, "Scenario end");
		} catch(final Exception e) {
			LogUtil.exception(LOG, Level.ERROR, e, "Scenario failed");
		}
	}
}
